### PR TITLE
[Inference API] Add Mistral Embeddings Support to Inference API

### PR DESCRIPTION
**Pull request by markjhoy**
This PR adds support for Mistral text embeddings (`text_embedding` task type) integration into the Inference API. 

### Prerequisites to Model Creation
* A Mistral Account on [La Plateforme](https://console.mistral.ai/)
* An API key generated for your account

### Model Creation:
```
PUT _inference/text_embedding/{model_id}
{
  "service": "mistral",
  "service_settings": {
    "api_key": "{api_key}",
    "model": “{mistral_embedding_model}”,
    "max_input_tokens": {max input tokens, default 8k},
  }
}
```

### Required Service Settings:
* `model`: [Mistral API models](https://docs.mistral.ai/getting-started/models/#operation/listModels) - only `mistral-embed` is available currently
* `api_key`: The API key can be found on your Azure AI Studio deployment's overview page

### Optional Service Settings
* `max_input_tokens`: (optional) the number of input token to use for truncation

### Task Settings
(no task settings available)

### Text Embedding Inference

```
POST _inference/text_embedding/{model_id}
{
    "input": "The answer to the universe is"
}
```

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        Integer maxInputTokens = extractOptionalPositiveInteger(
+            map,
+            MAX_INPUT_TOKENS,
+            ModelConfigurations.SERVICE_SETTINGS,
+            validationException
+        );
+        RateLimitSettings rateLimitSettings = RateLimitSettings.of(map, DEFAULT_RATE_LIMIT_SETTINGS, validationException);
+
+        Integer dims = removeAsType(map, DIMENSIONS, Integer.class);
+        Boolean dimensionsSetByUser = extractOptionalBoolean(map, DIMENSIONS_SET_BY_USER, validationException);
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:68]: It doesn't look like we use `dimensionsSetByUser` in the actual request to mistral. Do we need it? I think we'd only need it if we need to send the `dimensions` field to mistral. I also don't see it in their docs: https://docs.mistral.ai/api/#operation/createEmbedding

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return ADD_MISTRAL_EMBEDDINGS_INFERENCE;
+    }
+
+    public String model() {
+        return this.model;
+    }
+
+    public Integer dimensions() {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:134]: Let's add the `@Override` here

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    public boolean dimensionsSetByUser() {
+        return this.dimensionsSetByUser;
+    }
+
+    public Integer maxInputTokens() {
+        return this.maxInputTokens;
+    }
+
+    public SimilarityMeasure similarity() {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:146]: Let's add the `@Override` here

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        return this.similarity;
+    }
+
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:149]: We also need to override the elementType method:


```
    @Override
    public DenseVectorFieldMapper.ElementType elementType() {
        return DenseVectorFieldMapper.ElementType.FLOAT;
    }
```

We really need to make it not default to null 🤦‍♂️ too easy to forget.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralService.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        InputType inputType,
+        ChunkingOptions chunkingOptions,
+        TimeValue timeout,
+        ActionListener<List<ChunkedInferenceServiceResults>> listener
+    ) {
+        ActionListener<InferenceServiceResults> inferListener = listener.delegateFailureAndWrap(
+            (delegate, response) -> delegate.onResponse(translateToChunkedResults(input, response))
+        );
+
+        doInfer(model, input, taskSettings, inputType, timeout, inferListener);
```
[the remark this answers is not in the document]
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralService.java:106]: Ah - I'll take a look... TBH - I copied this part from the Azure AI Studio pieces :)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java (hunk 1 above)
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:68]: If it's not needed, that's great... I thought it was needed however for our processing on the results? (I'm probably wrong...)

Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:68]: Yeah I think we only need it if we have to send the `dimensions` field to mistral. We use it in the service but that's only to verify that we got the right number of dimensions back when we send it to the external provider.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralService.java:106]: I've added the chunking code to AI studio and other services were it was not implemented in #109050

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralConstants.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+public class MistralConstants {
+    public static final String MISTRAL_API_EMBEDDINGS_PATH = "https://api.mistral.ai/v1/embeddings";
+
+    public static final String MISTRAL_API_KEY_FIELD = "api_key";
+    public static final String DIMENSIONS_SET_BY_USER = "dimensions_set_by_user";
+
+    public static final String MISTRAL_MODEL_FIELD = "model";
+    public static final String MISTRAL_INPUT_FIELD = "input";
+    public static final String MISTRAL_ENCODING_FORMAT_FIELD = "encoding_format";
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralConstants.java:18]: super small nit: do we need the `MISTRAL_` prefix, if the class name already specifies `Mistral`?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsModel.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        Map<String, Object> taskSettings,
+        @Nullable Map<String, Object> secrets,
+        ConfigurationParseContext context
+    ) {
+        this(
+            inferenceEntityId,
+            taskType,
+            service,
+            MistralEmbeddingsServiceSettings.fromMap(serviceSettings, context),
+            new EmptyTaskSettings(),    // no task settings for Mistral embeddings
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsModel.java:49]:
```suggestion
            EmptyTaskSettings.INSTANCE,    // no task settings for Mistral embeddings
```

nit: I think we can reuse the same instance here without creating a new object

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        return this.rateLimitSettings;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(model);
+        out.writeOptionalVInt(dimensions);
+        out.writeBoolean(dimensionsSetByUser);
+        out.writeOptionalEnum(SimilarityMeasure.translateSimilarity(similarity, out.getTransportVersion()));
+        out.writeOptionalInt(maxInputTokens);
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:160]: Should this also be `writeOptionalVInt`?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+        return new MistralEmbeddingsServiceSettings(model, dims, dimensionsSetByUser, maxInputTokens, similarity, rateLimitSettings);
+    }
+
+    public MistralEmbeddingsServiceSettings(StreamInput in) throws IOException {
+        this.model = in.readString();
+        this.dimensions = in.readOptionalVInt();
+        this.dimensionsSetByUser = in.readBoolean();
+        this.similarity = in.readOptionalEnum(SimilarityMeasure.class);
+        this.maxInputTokens = in.readOptionalInt();
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsServiceSettings.java:100]: Should this also use `readOptionalVInt`?

**Review by timgrein — COMMENTED** [review]
Left some smaller comments, rest looks good :)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralConstants.java (hunk 6 above)
Comment (markjhoy) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralConstants.java:18]: Good call -- I've been way to deep in the weeds recently in these providers that I think I intuitively wanted to ensure they have unique names where there are other fields that are similar for other providers... :)

**Review by jonathan-buttner — APPROVED** [review]
Thanks for the changes!

**Comment by markjhoy** [comment]
> Thanks for the changes!

Thank you, and @timgrein for the great reviews as always 🎉

**Pull request merged**